README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# chantrics

<!-- badges: start -->

![R-CMD-check](https://github.com/tbruckbauer/chantrics/workflows/R-CMD-check/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/tbruckbauer/chantrics/branch/master/graph/badge.svg?token=EQQ177WODM)](https://codecov.io/gh/tbruckbauer/chantrics)
![pkgdown](https://github.com/tbruckbauer/chantrics/workflows/pkgdown/badge.svg)
[![CodeFactor](https://www.codefactor.io/repository/github/tbruckbauer/chantrics/badge?s=8292aacb6a947f6280b202c4f29e21e4510dce53)](https://www.codefactor.io/repository/github/tbruckbauer/chantrics)
<!-- badges: end -->

`chantrics` adjusts the loglikelihood of common econometric models for
clustered data based on the estimation process suggested in Chandler and
Bate (2007), using the
[chandwich](https://cran.r-project.org/package=chandwich) package, and
provides convenience functions for inference on the adjusted models.
`adj_loglik()` adjusts the parameter covariance matrix of the models to
incorporate clustered data, and can mitigate model misspecification by
wrapping `chandwich::adjust_loglik` for the supported models.

The returned model of class `chantrics` can be plugged into standard
model evaluation and model comparison methods, for example `summary()`,
`confint()` and `anova()`, and a hypothesis test framework provided by
`alrtest()`.

## Installation

You can install the released version of chantrics from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("chantrics")
```

The development version is available from [GitHub](https://github.com/)
with:

``` r
## Remove the # if "devtools" is not installed yet.
# install.packages("devtools")
devtools::install_github("tbruckbauer/chantrics")
```

## Usage

This example is using the misspecified count data regression from
Chapter 5.1 in the Object-Oriented Computation of Sandwich Estimators
vignette from the *sandwich* package (Zeileis 2006). First, data from a
negative binomial model is generated, and then a Poisson model is fit,
which is clearly misspecified.

``` r
library(chantrics)

set.seed(123)
x <- rnorm(250)
y <- rnbinom(250, mu = exp(1 + x), size = 1)

## Fit the Poisson glm model, which is not correctly specified
fm_pois <- glm(y ~ x + I(x ^ 2), family = poisson)
lmtest::coeftest(fm_pois)
#> 
#> z test of coefficients:
#> 
#>              Estimate Std. Error z value Pr(>|z|)    
#> (Intercept)  1.063268   0.041357 25.7094  < 2e-16 ***
#> x            0.996072   0.053534 18.6062  < 2e-16 ***
#> I(x^2)      -0.049124   0.023146 -2.1223  0.03381 *  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# The I(x^2) term is spuriously significant.
```

In order to mitigate the misspecification, the loglikelihood adjustment
is applied to the model object using `adj_loglik(model_object)`. If
clustered data is available, a vector or factor indicating from which
cluster the observation originates can be passed into the function using
`cluster`. If it is not supplied, it is assumed that each observation
originates from its own cluster and is independent. More information on
this can be found in the [clustering
vignette](https://chantrics.theobruckbauer.eu/articles/clustering-vignette.html).

``` r
## Apply the loglikelihood adjustment to the model
fm_pois_adj <- adj_loglik(fm_pois)
summary(fm_pois_adj)
#>                  MLE      SE adj. SE
#> (Intercept)  1.06300 0.04136 0.08378
#> x            0.99610 0.05353 0.10520
#> I(x^2)      -0.04912 0.02315 0.03628
lmtest::coeftest(fm_pois_adj)
#> 
#> z test of coefficients:
#> 
#>              Estimate Std. Error z value Pr(>|z|)    
#> (Intercept)  1.063268   0.083776 12.6918   <2e-16 ***
#> x            0.996072   0.105217  9.4668   <2e-16 ***
#> I(x^2)      -0.049124   0.036284 -1.3539   0.1758    
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# The I(x^2) term is no longer significant.

# Produce confidence intervals for the parameter estimates
chandwich::conf_intervals(fm_pois_adj)
#> Waiting for profiling to be done...
#> Model: poisson_glm_lm 
#> 
#> 95% confidence intervals, adjusted loglikelihod with type = ''vertical'' 
#> 
#> Symmetric:
#>              lower     upper   
#> (Intercept)   0.89907   1.22747
#> x             0.78985   1.20229
#> I(x^2)       -0.12024   0.02199
#> 
#> Profile likelihood-based:
#>              lower    upper  
#> (Intercept)   0.8954   1.2232
#> x             0.7877   1.1991
#> I(x^2)       -0.1198   0.0222
# Generate a nested model from fm_pois_adj
fm_pois_small_adj <- update(fm_pois_adj, . ~ . - I(x^2))
lmtest::coeftest(fm_pois_small_adj)
#> 
#> z test of coefficients:
#> 
#>             Estimate Std. Error z value  Pr(>|z|)    
#> (Intercept)  1.05525    0.08111  13.010 < 2.2e-16 ***
#> x            0.90172    0.06939  12.995 < 2.2e-16 ***
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# Perform an adjusted likelihood ratio test on the two models
anova(fm_pois_adj, fm_pois_small_adj)
#> Analysis of Adjusted Deviance Table
#> 
#> Model 1: y ~ x + I(x^2)
#> Model 2: y ~ x
#> 
#>   Resid.df df  ALRTS Pr(>ALRTS)
#> 1      247                     
#> 2      248  1 1.8202     0.1773
# Similarly, specify to remove I(x^2) from the larger model
alrtest(fm_pois_adj, "I(x^2)")
#> Adjusted likelihood ratio test
#> 
#> Model 1: y ~ x + I(x^2)
#> Model 2: y ~ x
#> 
#>   Resid.df df  ALRTS Pr(>ALRTS)
#> 1      247                     
#> 2      248  1 1.8202     0.1773
# Plot confidence regions for the parameter estimates
fm_pois_adj_vert <-
  chandwich::conf_region(fm_pois_adj, which_pars = c("x", "I(x^2)"))
#> Waiting for profiling to be done...
fm_pois_adj_none <-
  chandwich::conf_region(fm_pois_adj,
    which_pars = c("x", "I(x^2)"),
    type = "none"
  )
#> Waiting for profiling to be done...
plot(
  fm_pois_adj_vert,
  fm_pois_adj_none,
  conf = c(60, 80, 90, 95),
  col = c("brown", "darkgreen"),
  lty = c(1, 2),
  lwd = 2.5
)
```

<img src="man/figures/README-demo-1.png" width="100%" />

More information on the different methods can be found in the
[Introducing chantrics
vignette](https://chantrics.theobruckbauer.eu/articles/chantrics-vignette.html),
and in the corresponding help pages.

## References

<div id="refs" class="references csl-bib-body hanging-indent">

<div id="ref-chanbate07" class="csl-entry">

Chandler, Richard E., and Steven Bate. 2007. “Inference for Clustered
Data Using the Independence Loglikelihood.” *Biometrika* 94 (1): 167–83.
<https://doi.org/doi:10.1093/biomet/asm015>.

</div>

<div id="ref-zeileis06" class="csl-entry">

Zeileis, Achim. 2006. “Object-Oriented Computation of Sandwich
Estimators.” *Journal of Statistical Software, Articles* 16 (9): 1–16.
<https://doi.org/10.18637/jss.v016.i09>.

</div>

</div>


NEWS.md
# NEWS

## chantrics 1.0.0

First public release of the package.


R/glm.R
#' Loglikelihood adjustments for glm fits
#'
#' In a generalised linear model (glm), the user can choose between a range of
#' distributions of a response \eqn{y}, and can allow for non-linear relations
#' between the mean outcome for a particular combination of covariates \eqn{x},
#' \eqn{E(y_i\mid x_i)=\mu_i}, and the linear predictor,
#' \eqn{\eta_i=x_i^T\beta}, which is the link function \eqn{g(\mu_i)=\eta_i}.
#' it is required to be monotonic. (For a quick introduction, see
#' Kleiber and Zeileis (2008, Ch. 5.1), for more complete coverage of the
#' topic, see, for example, Davison (2003, Ch. 10.3))
#'
#' For more usage examples and more information on `glm` models, see the
#' *Introducing `chantrics`* vignette by running
#' \code{vignette("chantrics-vignette", package = "chantrics")}
#'
#' @section Supported families (within each family, any link function should work):
#'
#'   * `gaussian`
#'   * `poisson`
#'   * `binomial`
#'   * `MASS::negative.binomial`
#'
#'   Also works for [MASS::glm.nb()], note that the standard errors of the theta
#'   are not adjusted.
#'
#' @examples
#' # binomial example from Applied Econometrics in R, Kleiber/Zeileis (2008)
#' # ==  probit  ==
#' data("SwissLabor", package = "AER")
#' swiss_probit <- glm(participation ~ . + I(age^2),
#'   data = SwissLabor,
#'   family = binomial(link = "probit")
#' )
#' summary(swiss_probit)
#' swiss_probit_adj <- adj_loglik(swiss_probit)
#' summary(swiss_probit_adj)
#'
#' # == logit ==
#' swiss_logit <- glm(participation ~ . + I(age^2),
#'   data = SwissLabor,
#'   family = binomial(link = "logit")
#' )
#' summary(swiss_logit)
#' swiss_logit_adj <- adj_loglik(swiss_logit)
#' summary(swiss_logit_adj)
#' @references Davison, A. C. 2003. Statistical Models. Cambridge Series on
#'   Statistical and Probabilistic Mathematics 11. Cambridge University Press,
#'   Cambridge.
#'
#'   Kleiber, Christian, and Achim
#'   Zeileis. 2008. Applied Econometrics with R. Edited by Robert Gentleman,
#'   Kurt Hornik, and Giovanni Parmigiani. Use r! New York: Springer-Verlag.
#' @name glm
#' @aliases glm.nb
NULL

#' @export

# handling of dispersion parameters:
# http://people.stat.sfu.ca/~raltman/stat402/402L25.pdf

logLik_vec.glm <- function(object, pars = NULL, ...) {
  if (!missing(...)) {
    rlang::warn("extra arguments discarded")
  }
  # import coefficients
  if (is.null(pars)) {
    pars <- stats::coef(object)
  }
  # calculate mus
  # try getting the design matrix from glm object
  x_mat <- get_design_matrix_from_model(object)
  response_vec <- get_response_from_model(object)
  df.resid <- df.residual(object)
  # theta for constant negbin models
  theta <- try(object$call$family$theta, silent = TRUE)

  if (inherits(object, "negbin")) {
    family <- "negbin"
    # get theta if saved
    if (!is.numeric(theta)) {
      theta <- rlang::env_get(bypasses.env, "theta", default = NULL)
      if (!is.numeric(theta)) {
        theta <- rlang::env_get(bypasses.env, paste0("theta", length(stats::coef(object))), default = NULL)
      }
    }
  } else {
    family <- object$family$family
  }
  llv <- glm_type_llv(
    family = family,
    x_mat = x_mat,
    pars = pars,
    response_vec = response_vec,
    linkinv = object$family$linkinv,
    df.resid = df.resid,
    theta = theta
  )

  # return other attributes from logLik objects
  if (inherits(object, "negbin")) {
    pars <- c(pars, "theta")
  }
  attr(llv, "df") <- length(pars)
  attr(llv, "nobs") <- stats::nobs(object)
  class(llv) <- "logLik_vec"
  return(llv)
}

glm_type_llv <- function(family, x_mat, pars, response_vec, linkinv, df.resid = NULL, theta = NULL, hurdle = c("no", "count", "zero")) {
  hurdle <- match.arg(hurdle)
  linkinv <- match.fun(linkinv)
  dimcheck(x_mat, pars)
  eta_vec <- x_mat %*% pars
  mu_vec <- linkinv(eta_vec)
  # "geometric" is negbin with theta = 1
  if (family == "geometric") {
    theta <- 1
    family <- "Negative Binomial("
  }
  if (family == "poisson") {
    llv <- stats::dpois(response_vec, lambda = mu_vec, log = TRUE)
  } else if (family == "binomial") {
    llv <- stats::dbinom(response_vec, 1, prob = mu_vec, log = TRUE)
  } else if (family == "gaussian") {
    # estimate the dispersion parameter
    disp <- dispersion.gauss(response_vec, mu_vec, df.resid - 1)
    pars <- c(pars, disp)
    names(pars)[length(pars)] <- "Dispersion"
    llv <- stats::dnorm(response_vec - mu_vec, mean = 0, sd = sqrt(disp), log = TRUE)
  } else if (substr(family, 1, 18) == "Negative Binomial(") {
    llv <- stats::dnbinom(response_vec, size = theta, mu = mu_vec, log = TRUE)
  } else if (family == "negbin") {
    # if theta should not be estimated (if through confint, etc.), then this
    # should be set previously.
    # If theta is not a number, reestimate
    if (!is.numeric(theta)) {
      theta <- MASS::theta.ml(y = response_vec, mu = mu_vec)
    }
    llv <- stats::dnbinom(response_vec, size = theta, mu = mu_vec, log = TRUE)
    rlang::env_poke(bypasses.env, "last_est_theta", theta)
  } else {
    rlang::abort(paste0(family, " is not supported."), class = "chantrics_not_supported_glm_family")
  }
  if (hurdle == "count") {
    # subtract the probability of the function being equal to 0,
    # see countregression-cam, Eqn. 4.54
    if (any(family == "negbin", family == "Negative Binomial(")) {
      llv <- llv - log(1 - stats::dnbinom(response_vec, size = theta, mu = mu_vec, log = FALSE))
    } else if (family == "poisson") {
      llv <- llv - log(1 - stats::dpois(response_vec, lambda = mu_vec, log = FALSE))
    } else {
      # safety break
      rlang::abort("Could not match count distribution specification.")
    }
    # set all contributions whose count is 0 to 0
    llv[response_vec == 0] <- 0
  }
  return(llv)
}

#' @keywords internal

dispersion.gauss <- function(response_vec, mu_vec, df) {
  # https://statmath.wu.ac.at/courses/heather_turner/glmCourse_001.pdf
  # (last slide in GLM > Estimation section)

  # http://people.stat.sfu.ca/~raltman/stat402/402L25.pdf (p. 4) -> dispersion
  # is the residual variance
  # final source: ISL, page 80, eqn. 3.25 (RSE formula)
  # response_vec are the true Y, the mu_vec are the fitted values

  return(sum((response_vec - mu_vec)^2) / (df))
}

#' @keywords internal

dispersion.stat <- function(response_vec, mu_vec, object) {
  # modelcount-hilbe, pg. 78 (Eqn. 3.4, pearson chi-sq statistic), pg. 79,
  # "The dispersion statistic of the Poisson model is defined as the
  # Pearson Chi2 statistic divided by the residual degrees of freedom"
  return(sum(((response_vec - mu_vec)^2) / object$family$variance(mu_vec)) / (stats::df.residual(object)))
}

#' @keywords internal

dimcheck <- function(x_mat, pars) {
  # check that dimensions of xmat and pars coincide
  if (ncol(x_mat) != length(pars)) {
    rlang::abort(paste0(
      "The length of 'pars' (", length(pars), ") does not fit\n",
      "the number of columns in the design matrix, ", ncol(x_mat)
    ),
    class = "chantrics_pars_wrong_length"
    )
  }
  if (any(colnames(x_mat) != names(pars))) {
    rlang::warn(paste0(
      "names(pars) is not equal to the colnames() of the design matrix.\n",
      "Continuing with the parameters as matched below:",
      "Design matrix: ", paste(colnames(x_mat), sep = ", "),
      "\n names(pars): ", paste(names(pars), sep = ", ")
    ),
    class = "chantrics_parnames_do_not_match"
    )
  }
}

#' @keywords internal

fittedhelper.glm <- function(object, type) {
  x_mat <- get_design_matrix_from_model(object)
  pars <- stats::coef(object)
  result <- x_mat %*% pars
  if (type == "response") {
    result <- get_unadj_object(object)$family$linkinv(result)
  }
  dim(result) <- NULL
  names(result) <- seq(1, length(result), 1)
  return(result)
}
